apps/nextra/components/remote-codeblock/utils.ts

```diff
@@ -0,0 +1,207 @@
+  // If there's no dot, return an empty string (or adjust as needed)
+  if (lastDotIndex !== -1 && lastDotIndex + 1 < lastSegment.length) {
+    return lastSegment.substring(lastDotIndex + 1);
+  } else {
+    return "";
+  }
+}
+
+/**
+ * Reformat Permalink Response
+ *
+ * Given response.data, reformats an array of codeSnippets to include
+ * highlight information for shiki rendering and restructures response
+ * for use with SSG components implementing useData() in .mdx files
+ * e.g., <RemoteCodeblock />
+ */
+export async function reformatPermalinkResponse(codeSnippets: CodeSnippet[]) {
+  const formattedJsonResponse: CodecacheSSGProps<ParsedCodeSnippet> = {
+    props: { ssg: {} },
+  };
+
+  for (let value of codeSnippets) {
+    const extension = getFileExtension(value.github_permalink);
+    let language: BundledLanguage;
+    if (Object.keys(EXTENSION_TO_LANGUAGE).includes(extension)) {
+      language = EXTENSION_TO_LANGUAGE[extension];
+    } else {
+      language = EXTENSION_TO_LANGUAGE["bash"];
+    }
+    const highlightedSnippet = await highlightCode(value.code, language);
+    formattedJsonResponse.props.ssg[value.github_permalink] = {
+      ...value,
+      highlightedCode: highlightedSnippet,
+      language,
+    };
+  }
+
+  return formattedJsonResponse;
+}
+
+/**
+ * Fetches Github Permalinks from CodeCache
+ *
+ * Requires the existence of a NEXT_PUBLIC_CODECACHE_API_KEY in .env
+ * NOT RATE LIMITED, REQUIRES API KEY FOR CACHE
+ */
+export async function permalinkFetchFromCache(
+  permalinks: string[],
+): Promise<CodecacheSSGProps<ParsedCodeSnippet>> {
+  const apiKey = process.env.NEXT_PUBLIC_CODECACHE_API_KEY;
+  if (!apiKey) {
+    // Should panic
+    throw new Error(
+      "Unexpected error: fetching permalinks from cache but no NEXT_PUBLIC_CODECACHE_API_KEY defined in .env",
+    );
+  }
+
+  try {
+    const body = JSON.stringify({
+      github_permalinks: permalinks,
+    });
+
+    // Request all links from cache
+    const response = await fetch(`${baseRoute}/codecache`, {
+      headers: {
+        Origin,
+        "x-api-key": apiKey,
+      },
+      method: "POST",
+      body,
+    });
+
+    const jsonResponse: CodecacheResponse<CodeSnippet[]> =
+      await response.json();
+    const formattedJsonResponse: CodecacheSSGProps<ParsedCodeSnippet> =
+      await reformatPermalinkResponse(jsonResponse.data);
+    return formattedJsonResponse;
+  } catch (err) {
+    const errorMessage =
+      "An error occurred while fetching github permalinks for Remote Codeblock from codecache: ";
+    console.warn(errorMessage, err);
+    throw new Error(errorMessage, err);
+  }
+}
+
+/**
+ * Fetches Github Permalinks from Github
+ *
+ * Fallback in the event that NEXT_PUBLIC_CODECACHE_API_KEY is not set in .env
+ * NOTE: RATE LIMITED BY GITHUB, UP TO 5000 REQUESTS PER HOUR
+ *
+ * @see https://stackoverflow.com/questions/13394077/is-there-a-way-to-increase-the-api-rate-limit-or-to-bypass-it-altogether-for-git
+ */
```

Comment:
We do cache this clientside right?

Like if I go to a different page and come back

Reply:
No the way it works is I made it so that the Remote Codeblocks make requests during the site build so the code snippets are effectively cached and stored within the build folder as a part of the markdown content for that page. Thus when people visit the production site, no requests to Github api are made since the site content -- and all Codeblocks -- are statically built.

The tradeoff is that in development, when someone is editing a specific page, it will request the code snippets from the Github API for that specific page if they don't have the API Key. I looked into caching for `getStaticProps` in development but it doesn't seem to be available atm, but I think most folks will not come close to the rate limit during development so its fine.

Reply:
Yeah, if we start running into it.  We can just cache on disk or something for local development purposes.  good work